test/integration/apiserver/flowcontrol/concurrency_test.go

```diff
@@ -0,0 +1,245 @@
+
+	flowcontrolv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/pkg/master"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	dispatchedRequestCountMetricsName               = "apiserver_flowcontrol_dispatched_requests_total"
+	dispatchedRequestCountMetricsLabelPriorityLevel = "priorityLevel"
+	timeout                                         = time.Second * 10
+)
+
+func setup(t testing.TB) (*httptest.Server, *rest.Config, framework.CloseFunc) {
+	opts := framework.MasterConfigOptions{EtcdOptions: framework.DefaultEtcdOptions()}
+	opts.EtcdOptions.DefaultStorageMediaType = "application/vnd.kubernetes.protobuf"
+	masterConfig := framework.NewIntegrationTestMasterConfigWithOptions(&opts)
+	resourceConfig := master.DefaultAPIResourceConfigSource()
+	resourceConfig.EnableVersions(schema.GroupVersion{
+		Group:   "flowcontrol.apiserver.k8s.io",
+		Version: "v1alpha1",
+	})
+	masterConfig.GenericConfig.MaxRequestsInFlight = 5
+	masterConfig.GenericConfig.MaxMutatingRequestsInFlight = 5
+	masterConfig.GenericConfig.OpenAPIConfig = framework.DefaultOpenAPIConfig()
+	masterConfig.ExtraConfig.APIResourceConfigSource = resourceConfig
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+
+	return s, masterConfig.GenericConfig.LoopbackClientConfig, closeFn
+}
+
+func TestPriorityLevelIsolation(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, genericfeatures.APIPriorityAndFairness, true)()
+	// NOTE: disabling the feature should fail the test
+	_, loopbackConfig, closeFn := setup(t)
+	defer closeFn()
+
+	loopbackClient := clientset.NewForConfigOrDie(loopbackConfig)
+	noxu1Client := getClientFor(loopbackConfig, "noxu1")
+	noxu2Client := getClientFor(loopbackConfig, "noxu2")
+
+	priorityLevelNoxu1, _, err := createPriorityLevelAndBindingFlowSchemaForUser(loopbackClient, "noxu1")
+	require.NoError(t, err)
+	priorityLevelNoxu2, _, err := createPriorityLevelAndBindingFlowSchemaForUser(loopbackClient, "noxu2")
+	require.NoError(t, err)
+
+	wg := &sync.WaitGroup{}
+	// "elephant"
+	streamRequests(wg, 20, 100, func() {
+		_, err := noxu1Client.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
+		require.NoError(t, err)
+	})
+
+	streamRequests(nil, 1, 100, func() {
+		_, err := noxu2Client.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
+		require.NoError(t, err)
+	})
+
+	wg.Wait()
+
+	reqCounts, err := getRequestCountOfPriorityLevel(loopbackClient)
+
+	assert.Equal(t, 2000, reqCounts[priorityLevelNoxu1.Name])
+	assert.Equal(t, 100, reqCounts[priorityLevelNoxu2.Name])
```

Comment:
Isn't the ratio wrong now? The elephant is sending 20x the requests but only has 10x the concurrency shares, and it's the latter that matters?

I think this check for exactly 100 is going to be flaky. By my math, if we desire a flake rate of < 1/10000 ("4 nines"), and the time take of an individual namespace list call varies by 5% (including e.g. CPU scheduling noise, and also OS process scheduling noise, since there's a real etcd for integration tests), then the check should be >= 89. ...but I doubt namespace list calls vary by that little. If they vary by 20%, then we can basically say nothing about how many we expect to have happened and be wrong less than 1 in 10,000 times. (I could well be wrong, but I believe 100 is a large enough sample size that the central limit theorem applies and you can use the table here to do the math: https://en.wikipedia.org/wiki/Normal_distribution#Standard_deviation_and_coverage)

Reply:
@lavalamp:

- the elephant and mouse have the same number of concurrency shares and so will get the same concurrency limit.

- the mouse can be expected to have 100 successful calls because they come serially from a single client goroutine and get assigned to their own private priority level with a non-zero concurrency limit.  It is the elephant that I fear will flake, because it has 20 clients and a capacity of 11.

Reply:
The elephant can never flake, because the waitgroup waits for it to finish. Did I miss something?

Reply:
Oops, I missed that the mouse no longer uses the WaitGroup.  So yes there is a serious danger of the mouse flaking.

The elephant should fail because it has 20 concurrent clients but a server-side concurrency capacity of 11 or 15 (depending on whether the suggested config objects are injected).

Reply:
10 requests can be queued, and the requests execute quickly, so I think it probably will have all of its requests succeed.

The variable name here should be changed, since "dispatched" requests is going to be a different number than total requests, and I agree that's a source of flakiness.

Reply:
I am confused-- just because we make 20 goroutines does not mean that the requests actually attempt to execute concurrently? E.g. 10 execute, 10 queue, those 10 execute while the clients are getting the answer and sending the next 20, the cycle repeats.

Reply:
With a concurrency limit of 1 or 5, there can not be 10 executing concurrently.  That's why I say the server-side capacity is 11 or 15: 10 in the queue plus 1 or 5 more executing under the concurrency limit.

Reply:
There can't be 10 executing concurrently when the concurrency limit is 1 or 5.

Reply:
OK, understood. However, looking at the streamRequests function, I'm not convinced the requests are "really" in parallel.

Reply:
@lavalamp: I think you may be getting to my larger concerns about overheads and CPU scheduling fairness.  The code clearly launches 21 goroutines and the only synchronization is that after launching all 21 it waits for all of the first 20 to finish.  If the overheads are relatively small and the CPU scheduling is relatively fair then there will be 21 requests at the server concurrently.  If the existing test has executed successfully then that proves that either or both of those assumptions are false, but does not prove they will always be false.  That would mean this is a flaky test.

Reply:
Regardless of goroutines, I think everything is serialized by the act of making network connections.

3 phase model:
A) Request is sent to server (time for opening connection and transmitting the initial packet(s))
B) Server finishes request (time until request stops consuming a queue and/or concurrency slot)
C) Request completion (time until all response bytes are sent to client)

The first N (=21) requests all succeed as long as A < (5+10)*B, which is believable (since B involves its own trivial network request to etcd, I'd expect it to be a bit more expensive than A+C). Subsequent requests succeed as long as A+C < (5+10)*B, which is very believable (since a goroutine waits for the previous request's finish before starting the next request).

I think the test is flaky unless we can show that one of the factors above dominates the others by like 100 times or something. Ensuring B is very slow is one way. Making N drastically larger isn't as good because computers don't have infinite capacity to run goroutines in parallel. Reducing queue slots to e.g. 1 and effective concurrency to 1 or 2 might be enough to at least illustrate the problem.

Reply:
Phasing does not help if all the concurrent threads try to go through the same phase at the same time, which they will if enough real threads are available.  What happens if not enough real threads are available is a matter that we have no guarantees on.

I had forgotten that the client retries in response to a 429, up to 10 times.  It is quite plausible that this makes the current test work --- for either of two reasons.  One would be that the rejections are spread around the clients with a modicum of fairness, so no client racks up 10 consecutive rejections.  The other would be that it takes less than 10 wait-and-retry cycles to let the winners finish their 100 requests to list all the namespaces.